<a id="0"></a> <br>
### <span style='background :PowderBlue'>Índice</span>

1. [Carregamento e inspeção inicial](#1)
2. [Coluna `name`](#2)
3. [Coluna `author`](#3)
4. [Coluna `narrator`](#4)
5. [Coluna `time`](#5)
6. [Coluna `releasedate`](#6)
7. [Coluna `language`](#7)
8. [Coluna `stars`](#8)
9. [Coluna `price`](#9)
10. [Conclusão](#10)

<a id="1"></a>
### <span style='background :PowderBlue'>Carregamento e inspeção inicial</span>

In [1]:
# Importar as bibliotecas Pandas e NumPy.
import pandas as pd
import numpy as np

In [2]:
# Carregar o arquivo CSV no DataFrame df.
df = pd.read_csv('audible_uncleaned.csv')

# Inspecionar as primeiras 10 linhas do DataFrame.
df.head(10)

name                     author  \
0                  Geronimo Stilton #11 & #12  Writtenby:GeronimoStilton   
1                            The Burning Maze      Writtenby:RickRiordan   
2                                The Deep End       Writtenby:JeffKinney   
3                        Daughter of the Deep      Writtenby:RickRiordan   
4  The Lightning Thief: Percy Jackson, Book 1      Writtenby:RickRiordan   
5           The Hunger Games: Special Edition   Writtenby:SuzanneCollins   
6                 Quest for the Diamond Sword     Writtenby:WinterMorgan   
7                           The Dark Prophecy      Writtenby:RickRiordan   
8                   Merlin Mission Collection  Writtenby:MaryPopeOsborne   
9                           The Tyrant’s Tomb      Writtenby:RickRiordan   

                     narrator                time releasedate language  \
0       Narratedby:BillLobely   2 hrs and 20 mins    04-08-08  English   
1    Narratedby:RobbieDaymond   13 hrs and 8 mins    01-05-18  English   
2       Narratedby:DanRussell    2 hrs and 3 mins    06-11-20  English   
3   Narratedby:SoneelaNankani  11 hrs and 16 mins    05-10-21  English   
4   Narratedby:JesseBernstein              10 hrs    13-01-10  English   
5   Narratedby:TatianaMaslany  10 hrs and 35 mins    30-10-18  English   
6      Narratedby:LukeDaniels   2 hrs and 23 mins    25-11-14  English   
7    Narratedby:RobbieDaymond  12 hrs and 32 mins    02-05-17  English   
8  Narratedby:MaryPopeOsborne  10 hrs and 56 mins    02-05-17  English   
9    Narratedby:RobbieDaymond  13 hrs and 22 mins    24-09-19  English   

                           stars     price  
0     5 out of 5 stars34 ratings    468.00  
1   4.5 out of 5 stars41 ratings    820.00  
2   4.5 out of 5 stars38 ratings    410.00  
3   4.5 out of 5 stars12 ratings    615.00  
4  4.5 out of 5 stars181 ratings    820.00  
5     5 out of 5 stars72 ratings    656.00  
6     5 out of 5 stars11 ratings    233.00  
7     5 out of 5 stars50 ratings    820.00  
8      5 out of 5 stars5 ratings  1,256.00  
9     5 out of 5 stars58 ratings    820.00

O DataFrame possui as seguintes **colunas**:

* `name`: título do audiolivro.
* `author`: autor(es) do audiolivro.
* `narrator`: narrador(es) do audiolivro.
* `time`: duração do audiolivro, em horas e minutos.
* `releasedate`: data de lançamento do audiolivro, no formato dd-mm-aa.
* `language`: idioma do audiolivro.
* `stars`: média das avaliações dadas ao audiolivro, de 0 a 5, seguido do número de avaliações
* `price`: preço em Rupia indonésia.

In [3]:
# Inspecionar a presença de valores nulos, em que colunas ocorrem, e o tipo de dado de cada coluna.
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 87489 entries, 0 to 87488
Data columns (total 8 columns):
 #   Column       Non-Null Count  Dtype 
---  ------       --------------  ----- 
 0   name         87489 non-null  object
 1   author       87489 non-null  object
 2   narrator     87489 non-null  object
 3   time         87489 non-null  object
 4   releasedate  87489 non-null  object
 5   language     87489 non-null  object
 6   stars        87489 non-null  object
 7   price        87489 non-null  object
dtypes: object(8)
memory usage: 5.3+ MB


O DataFrame possui **87489** observações não-nulas por coluna de um total de 87489 observações. À primeira vista, **não há valores nulos** no DataFrame.

Nota-se que os **tipos de dados** de algumas colunas não são os mais adequados. Estes serão corrigidos ao longo do projeto, coluna a coluna. 

`time`, `stars` e `price` deverão ser alterados para um tipo de dado numérico, para facilitar sua manipulação. Será dado preferência ao tipo `float` por ser o tipo mais apropriado para análises de dados exploratórias que possam ser feitas sobre o DataFrame. 

`releasedate` será convertido para `datetime`, no formato dd-mm-aaaa, mais comum no Brasil. As demais colunas (`name`, `author`, `narrator`, `language`), por serem strings, permanecerão como tipo de dado `object`.

In [4]:
# Retornar True para linhas duplicadas.
duplicates = df.duplicated()

# Exibir o número de linhas duplicadas (True), se houver.
print(duplicates.value_counts())

False    87489
dtype: int64


**Não há linhas duplicadas** no DataFrame.

<a id="2"></a>
### <span style='background :PowderBlue'>Coluna `name`</span>

Serão removidos possíveis **espaços extras**, caso estes estejam presentes no DataFrame. Por se tratar de um DataFrame longo e com muitas observações únicas, não é factível avaliar coluna a coluna a necessidade de utilizar a função `.strip()`. Portanto, esta será utilizada em todas as colunas de maneira **preventiva**.

In [5]:
# Remover espaços extras.
df['name'] = df['name'].str.strip()

Inspecionando o DataFrame, foi possível notar que alguns títulos apresentam espaços antes de dois caracteres especiais: ? e !.

Portanto, serão **removidos os espaços antes de '?' e '!'**.

In [6]:
# Substituir ' ?' por '?', removendo o espaço extra.
df['name'] = df['name'].str.replace(' ?', '?', regex=False) # regex=False não irá considerar ? como uma expressão regular, e sim como caractere.

In [7]:
# Substituir ' !' por '!', removendo o espaço extra.
df['name'] = df['name'].str.replace(' !', '!')

# Inspecionar as primeiras 5 linhas do DataFrame até agora.
df.head()

name                     author  \
0                  Geronimo Stilton #11 & #12  Writtenby:GeronimoStilton   
1                            The Burning Maze      Writtenby:RickRiordan   
2                                The Deep End       Writtenby:JeffKinney   
3                        Daughter of the Deep      Writtenby:RickRiordan   
4  The Lightning Thief: Percy Jackson, Book 1      Writtenby:RickRiordan   

                    narrator                time releasedate language  \
0      Narratedby:BillLobely   2 hrs and 20 mins    04-08-08  English   
1   Narratedby:RobbieDaymond   13 hrs and 8 mins    01-05-18  English   
2      Narratedby:DanRussell    2 hrs and 3 mins    06-11-20  English   
3  Narratedby:SoneelaNankani  11 hrs and 16 mins    05-10-21  English   
4  Narratedby:JesseBernstein              10 hrs    13-01-10  English   

                           stars   price  
0     5 out of 5 stars34 ratings  468.00  
1   4.5 out of 5 stars41 ratings  820.00  
2   4.5 out of 5 stars38 ratings  410.00  
3   4.5 out of 5 stars12 ratings  615.00  
4  4.5 out of 5 stars181 ratings  820.00

Nota-se que nem todos os título estão capitalizados. Porém, **regras de capitalização possuem normas diferentes dependendo do idioma**. Como o DataFrame apresenta diversos idiomas, a capitalização dos títulos será mantida da maneira como está (a qual está correta para o inglês, idioma da maioria dos audiolivros presentes neste DataFrame).

In [8]:
# Inspecionar o número de audiolivros para cada linguagem do DataFrame.
df.language.value_counts()

English             61884
german               8295
spanish              3496
japanese             3167
italian              2694
french               2386
russian              1804
danish                935
portuguese            526
swedish               515
Hindi                 436
polish                224
finnish               197
dutch                 190
tamil                 161
catalan               153
mandarin_chinese       97
icelandic              52
romanian               50
hungarian              36
urdu                   34
afrikaans              28
czech                  23
turkish                20
greek                  18
arabic                 16
norwegian              16
galician               10
bulgarian               9
korean                  4
slovene                 4
hebrew                  2
basque                  2
telugu                  2
lithuanian              2
ukrainian               1
Name: language, dtype: int64

<a id="3"></a>
### <span style='background :PowderBlue'>Coluna `author`</span>

Remover **espaços extras**, se houverem.

In [9]:
# Remover espaços extras.
df['author'] = df['author'].str.strip()

**Remover 'Writtenby:'** para que a coluna contenha apenas o nome e sobrenome de cada autor.

In [10]:
# Remover "Writtenby:" substituindo-o por um caractere vazio.
df['author'] = df['author'].str.replace('Writtenby:', '')

Adicionar um **espaço entre nome e sobrenome**, os quais não tem espaçamento na coluna.

A linha de código abaixo irá modificar também autores com mais de um sobrenome, com sobrenomes abreviados, e audiolivros que possuem mais de um autor, adicionando espaçamento entre nome e sobrenome para todos.

In [11]:
# Substituir qualquer letra maiúscula por um espaço + a mesma letra utilizando expressões regulares.
df['author'] = df['author'].str.replace('([A-Z]+)', ' \\1', regex=True)

Pode-se notar que alguns audiolivros possuem **mais de um autor**.

A coluna `author` será dividida em 4 colunas, cada uma contendo o nome de um autor.

In [12]:
# Separar colunas por autor (author0, author1, author2, e author3) e adicioná-las ao DatFrame.
df = pd.concat([df['name'],
               df['author'].str.split(',', expand = True).add_prefix('author'),
               df.loc[:,['narrator', 'time', 'releasedate', 'language', 'stars', 'price']]], 
               axis = 1)

# Inspecionar as primeiras 5 linhas do DataFrame até agora.
df.head()

name            author0 author1  \
0                  Geronimo Stilton #11 & #12   Geronimo Stilton    None   
1                            The Burning Maze       Rick Riordan    None   
2                                The Deep End        Jeff Kinney    None   
3                        Daughter of the Deep       Rick Riordan    None   
4  The Lightning Thief: Percy Jackson, Book 1       Rick Riordan    None   

  author2 author3                   narrator                time releasedate  \
0    None    None      Narratedby:BillLobely   2 hrs and 20 mins    04-08-08   
1    None    None   Narratedby:RobbieDaymond   13 hrs and 8 mins    01-05-18   
2    None    None      Narratedby:DanRussell    2 hrs and 3 mins    06-11-20   
3    None    None  Narratedby:SoneelaNankani  11 hrs and 16 mins    05-10-21   
4    None    None  Narratedby:JesseBernstein              10 hrs    13-01-10   

  language                          stars   price  
0  English     5 out of 5 stars34 ratings  468.00  
1  English   4.5 out of 5 stars41 ratings  820.00  
2  English   4.5 out of 5 stars38 ratings  410.00  
3  English   4.5 out of 5 stars12 ratings  615.00  
4  English  4.5 out of 5 stars181 ratings  820.00

<a id="4"></a>
###  <span style='background :PowderBlue'>Coluna `narrator`</span>

**Remover espaços extras**, se houverem.

In [13]:
# Remover espaços extras.
df['narrator'] = df['narrator'].str.strip()

**Remover 'Narratedby:'** para que a coluna contenha apenas o nome e sobrenome de cada autor.

In [14]:
# Remover "Narratedby::" substituindo-o por um caractere vazio.
df['narrator'] = df['narrator'].str.replace('Narratedby:', '')

Assim como na coluna `author`, será adicionado um **espaço entre nome e sobrenome**.

Também serão modificados narradores com mais de um sobrenome, com sobrenomes abreviados, e audiolivros com mais de um narrador, se houverem.

In [15]:
# Substituir qualquer letra maiúscula por um espaço + a mesma letra utilizando expressões regulares.
df['narrator'] = df['narrator'].str.replace('([A-Z]+)', ' \\1', regex=True)

Como na coluna `author`, alguns audiolivros também possuem mais de um narrador.

A coluna `narrator` será dividida em 4 colunas, cada uma contendo o nome de um narrador.

In [16]:
# Separar colunas por narrador (narrator0, narrator1, narrator2, e narrator3) e adicioná-las ao DatFrame.
df = pd.concat([df['name'], df['author0'], df['author1'], df['author2'], df['author3'],
               df['narrator'].str.split(',', expand = True).add_prefix('narrator'),
               df.loc[:,['time', 'releasedate', 'language', 'stars', 'price']]], 
               axis = 1)

# Inspecionar as primeiras 5 linhas do DataFrame até agora.
df.head()

name            author0 author1  \
0                  Geronimo Stilton #11 & #12   Geronimo Stilton    None   
1                            The Burning Maze       Rick Riordan    None   
2                                The Deep End        Jeff Kinney    None   
3                        Daughter of the Deep       Rick Riordan    None   
4  The Lightning Thief: Percy Jackson, Book 1       Rick Riordan    None   

  author2 author3         narrator0 narrator1 narrator2 narrator3  \
0    None    None       Bill Lobely      None      None      None   
1    None    None    Robbie Daymond      None      None      None   
2    None    None       Dan Russell      None      None      None   
3    None    None   Soneela Nankani      None      None      None   
4    None    None   Jesse Bernstein      None      None      None   

                 time releasedate language                          stars  \
0   2 hrs and 20 mins    04-08-08  English     5 out of 5 stars34 ratings   
1   13 hrs and 8 mins    01-05-18  English   4.5 out of 5 stars41 ratings   
2    2 hrs and 3 mins    06-11-20  English   4.5 out of 5 stars38 ratings   
3  11 hrs and 16 mins    05-10-21  English   4.5 out of 5 stars12 ratings   
4              10 hrs    13-01-10  English  4.5 out of 5 stars181 ratings   

    price  
0  468.00  
1  820.00  
2  410.00  
3  615.00  
4  820.00

<a id="5"></a>
###  <span style='background :PowderBlue'>Coluna `time`</span>

**Remover espaços extras**, se houverem.

In [17]:
# Inspecionar as primeiras 5 linhas do DataFrame até agora.
df['time'] = df['time'].str.strip()

Para realizar a conversão da coluna `time` para um tipo de dado numérico, o que irá possibilitar a realização de análises numéricas sobre a coluna, iremos transformar os valores, atualmente guardados como 'x hours and x mins', para valores em **minutos**.

Serão extraídos apenas os dígitos correspondentes às horas, e guardados na nova coluna `hours`; o mesmo será feito para os minutos, que serão guardados na nova coluna `minutos`. Então, iremos converter os valores da coluna `hours` para minutos, e somá-los aos valores da coluna `minutos` para obter a duração total de cada audiolivro em minutos. Esta será guardada na coluna original `time`.

In [18]:
# Extrair um ou mais dígitos anteriores à string 'hr' ou 'hrs' e guardá-los na coluna 'hours' utilizando expressões regulares.
# Então, convertê-los para o tipo float para permitir manipulação numérica.
df['hours'] = df['time'].str.extract(r'(\d+) hr?s', expand=False).astype(float)

# Extrair um ou mais dígitos anteriores à string 'min' ou 'mins' e guardá-los na coluna 'hours' utilizando expressões regulares.
# Então, convertê-los para o tipo float para permitir manipulação numérica.
df['minutes'] = df['time'].str.extract(r'(\d+) min?s', expand=False).astype(float)

# Substituir valores nulos por 0 para realizar a conversão para minutos.
df['hours'].replace(np.nan, 0, inplace=True)
df['minutes'].replace(np.nan, 0, inplace=True)

# Substituir os valores na coluna 'time' pelos valores convertidos em minutos.
df['time'] = df['hours'] * 60 + df['minutes']

# Remover as colunas 'hours' e 'minutes', as quais não serão mais utilizadas.
df = df.drop(['hours', 'minutes'], axis=1)

# Inspecionar as primeiras 5 linhas do DataFrame até agora.
df.head()

name            author0 author1  \
0                  Geronimo Stilton #11 & #12   Geronimo Stilton    None   
1                            The Burning Maze       Rick Riordan    None   
2                                The Deep End        Jeff Kinney    None   
3                        Daughter of the Deep       Rick Riordan    None   
4  The Lightning Thief: Percy Jackson, Book 1       Rick Riordan    None   

  author2 author3         narrator0 narrator1 narrator2 narrator3   time  \
0    None    None       Bill Lobely      None      None      None  140.0   
1    None    None    Robbie Daymond      None      None      None  788.0   
2    None    None       Dan Russell      None      None      None  123.0   
3    None    None   Soneela Nankani      None      None      None  676.0   
4    None    None   Jesse Bernstein      None      None      None  600.0   

  releasedate language                          stars   price  
0    04-08-08  English     5 out of 5 stars34 ratings  468.00  
1    01-05-18  English   4.5 out of 5 stars41 ratings  820.00  
2    06-11-20  English   4.5 out of 5 stars38 ratings  410.00  
3    05-10-21  English   4.5 out of 5 stars12 ratings  615.00  
4    13-01-10  English  4.5 out of 5 stars181 ratings  820.00

<a id="6"></a>
###  <span style='background :PowderBlue'>Coluna `releasedate`</span>

**Remover espaços extras**, se houverem.

In [19]:
# Remover espaços extras.
df['releasedate'] = df['releasedate'].str.strip()

Os valores da coluna, atualmente guardados no formato dd-mm-aa, serão transformados para o **formato dd-mm-aaaa**, mais utilizado no Brasil. A coluna também será convertida para formato de data através da função `pd.to_datetime()`.

In [20]:
# Converter a coluna 'data' para um objetos Pandas datetime.
df['releasedate'] = pd.to_datetime(df['releasedate'], dayfirst=True) # Especificar que o dia virá primeiro.

# Converter as datas para o formato dd-mm-aaaa
df['releasedate'] = df['releasedate'].dt.strftime('%d-%m-%Y')

df.head()

name            author0 author1  \
0                  Geronimo Stilton #11 & #12   Geronimo Stilton    None   
1                            The Burning Maze       Rick Riordan    None   
2                                The Deep End        Jeff Kinney    None   
3                        Daughter of the Deep       Rick Riordan    None   
4  The Lightning Thief: Percy Jackson, Book 1       Rick Riordan    None   

  author2 author3         narrator0 narrator1 narrator2 narrator3   time  \
0    None    None       Bill Lobely      None      None      None  140.0   
1    None    None    Robbie Daymond      None      None      None  788.0   
2    None    None       Dan Russell      None      None      None  123.0   
3    None    None   Soneela Nankani      None      None      None  676.0   
4    None    None   Jesse Bernstein      None      None      None  600.0   

  releasedate language                          stars   price  
0  04-08-2008  English     5 out of 5 stars34 ratings  468.00  
1  01-05-2018  English   4.5 out of 5 stars41 ratings  820.00  
2  06-11-2020  English   4.5 out of 5 stars38 ratings  410.00  
3  05-10-2021  English   4.5 out of 5 stars12 ratings  615.00  
4  13-01-2010  English  4.5 out of 5 stars181 ratings  820.00

<a id="7"></a>
###  <span style='background :PowderBlue'>Coluna `language`</span>

**Remover espaços extras**, se houverem.

In [21]:
# Remover espaços extras.
df['language'] = df['language'].str.strip()

A maioria dos idiomas não estão capitalizados, como pode ser visto a seguir:

In [22]:
# Exibir todos os idiomas distintos do DataFrame.
df['language'].unique()

array(['English', 'Hindi', 'spanish', 'german', 'french', 'catalan',
       'swedish', 'italian', 'danish', 'finnish', 'dutch', 'hebrew',
       'russian', 'polish', 'galician', 'afrikaans', 'icelandic',
       'romanian', 'japanese', 'tamil', 'portuguese', 'urdu', 'hungarian',
       'czech', 'bulgarian', 'mandarin_chinese', 'basque', 'korean',
       'arabic', 'greek', 'turkish', 'ukrainian', 'slovene', 'norwegian',
       'telugu', 'lithuanian'], dtype=object)

Portanto, os idiomas serão **capitalizados** para padronizar a coluna.

In [23]:
# Capitalizar a coluna 'language'.
df['language'] = df['language'].str.title()

# Inspecionar as primeiras 5 linhas do DataFrame até agora.
df.head()

name            author0 author1  \
0                  Geronimo Stilton #11 & #12   Geronimo Stilton    None   
1                            The Burning Maze       Rick Riordan    None   
2                                The Deep End        Jeff Kinney    None   
3                        Daughter of the Deep       Rick Riordan    None   
4  The Lightning Thief: Percy Jackson, Book 1       Rick Riordan    None   

  author2 author3         narrator0 narrator1 narrator2 narrator3   time  \
0    None    None       Bill Lobely      None      None      None  140.0   
1    None    None    Robbie Daymond      None      None      None  788.0   
2    None    None       Dan Russell      None      None      None  123.0   
3    None    None   Soneela Nankani      None      None      None  676.0   
4    None    None   Jesse Bernstein      None      None      None  600.0   

  releasedate language                          stars   price  
0  04-08-2008  English     5 out of 5 stars34 ratings  468.00  
1  01-05-2018  English   4.5 out of 5 stars41 ratings  820.00  
2  06-11-2020  English   4.5 out of 5 stars38 ratings  410.00  
3  05-10-2021  English   4.5 out of 5 stars12 ratings  615.00  
4  13-01-2010  English  4.5 out of 5 stars181 ratings  820.00

<a id="8"></a>
###  <span style='background :PowderBlue'>Coluna `stars`</span>

**Remover espaços extras**, se houverem.

In [24]:
# Remover espaços extras.
df['stars'] = df['stars'].str.strip()

Nota-se que a coluna `stars` possui dois tipos de informação: a média das avaliações, e o número de avaliações recebidas pelo audiolivro.

Para permitir a realização de análises sobre os dois tipos de informação, estas serão **separadas em duas colunas**: `stars`, contendo a média das avaliações; e `ratings`, contendo o número de avaliações.

Primeiramente, o número de avaliações será extraído e guardado na nova coluna `ratings`.

In [25]:
# Extrair os dígitos imediatamente anteriores à string 'ratings', e convertê-los para o tipo float.
df['ratings'] = df['stars'].str.extract(r'(\d+) ratings', expand=False).astype(float)

# Substituir valores nulos por 0 para facilitar manipulação numérica.
df['ratings'].replace(np.nan, 0, inplace=True)

Então, será guardada apenas a média das avaliações na coluna `stars`.

In [26]:
# Extrair os primeiros dígitos da coluna 'stars' e convertê-los para o tipo float.
df['stars'] = df['stars'].str.extract('(^[0-9].[0-9]*)', expand=False).astype(float)

# Mudar a ordem das colunas para que 'ratings' venha logo após 'stars'.
df = df[['name', 'author0', 'author1', 'author2', 'author3', 'narrator0', 'narrator1', 'narrator2', 'narrator3', 'time', 
         'releasedate', 'language', 'stars', 'ratings', 'price']]

# Inspecionar as primeiras 5 linhas do DataFrame até agora.
df.head()

name            author0 author1  \
0                  Geronimo Stilton #11 & #12   Geronimo Stilton    None   
1                            The Burning Maze       Rick Riordan    None   
2                                The Deep End        Jeff Kinney    None   
3                        Daughter of the Deep       Rick Riordan    None   
4  The Lightning Thief: Percy Jackson, Book 1       Rick Riordan    None   

  author2 author3         narrator0 narrator1 narrator2 narrator3   time  \
0    None    None       Bill Lobely      None      None      None  140.0   
1    None    None    Robbie Daymond      None      None      None  788.0   
2    None    None       Dan Russell      None      None      None  123.0   
3    None    None   Soneela Nankani      None      None      None  676.0   
4    None    None   Jesse Bernstein      None      None      None  600.0   

  releasedate language  stars  ratings   price  
0  04-08-2008  English    5.0     34.0  468.00  
1  01-05-2018  English    4.5     41.0  820.00  
2  06-11-2020  English    4.5     38.0  410.00  
3  05-10-2021  English    4.5     12.0  615.00  
4  13-01-2010  English    4.5    181.0  820.00

<a id="9"></a>
###  <span style='background :PowderBlue'>Coluna `price`</span>

**Remover espaços extras**, se houverem.

In [27]:
# Remover espaços extras.
df['price'] = df['price'].str.strip()

Para possibilitar a realização de análises quantitativas sobre esta coluna, será preciso **convertê-la para um formato numérico**, como `float`.

Primeiramente, serão **removidas as vírgulas** para permitir a manipulação numérica da coluna utilizando Python. Além disso, as linhas que possuem como valor `Free` serão **substituídas por 0**.

In [28]:
# Retirar vírgulas para facilitar manipulação numérica.
df['price'] = df['price'].str.replace(',', '')

# Substituir 'Free' por 0 para permitir manipulação numérica.
df['price'] = df['price'].str.replace('Free', '0')

Então, será feita a conversão da coluna para `float`.

In [29]:
# Converter os valores de 'price' para float.
df['price'] = pd.to_numeric(df['price'])

# Inspecionar as primeiras 5 linhas do DataFrame até agora.
df.head()

name            author0 author1  \
0                  Geronimo Stilton #11 & #12   Geronimo Stilton    None   
1                            The Burning Maze       Rick Riordan    None   
2                                The Deep End        Jeff Kinney    None   
3                        Daughter of the Deep       Rick Riordan    None   
4  The Lightning Thief: Percy Jackson, Book 1       Rick Riordan    None   

  author2 author3         narrator0 narrator1 narrator2 narrator3   time  \
0    None    None       Bill Lobely      None      None      None  140.0   
1    None    None    Robbie Daymond      None      None      None  788.0   
2    None    None       Dan Russell      None      None      None  123.0   
3    None    None   Soneela Nankani      None      None      None  676.0   
4    None    None   Jesse Bernstein      None      None      None  600.0   

  releasedate language  stars  ratings  price  
0  04-08-2008  English    5.0     34.0  468.0  
1  01-05-2018  English    4.5     41.0  820.0  
2  06-11-2020  English    4.5     38.0  410.0  
3  05-10-2021  English    4.5     12.0  615.0  
4  13-01-2010  English    4.5    181.0  820.0

<a id="10"></a>
###  <span style='background :PowderBlue'>Conclusão</span>

Serão exibidas as cinco primeiras linhas do DataFrame limpo para melhor observar as operações realizadas durante a limpeza.

In [30]:
df.head()

name            author0 author1  \
0                  Geronimo Stilton #11 & #12   Geronimo Stilton    None   
1                            The Burning Maze       Rick Riordan    None   
2                                The Deep End        Jeff Kinney    None   
3                        Daughter of the Deep       Rick Riordan    None   
4  The Lightning Thief: Percy Jackson, Book 1       Rick Riordan    None   

  author2 author3         narrator0 narrator1 narrator2 narrator3   time  \
0    None    None       Bill Lobely      None      None      None  140.0   
1    None    None    Robbie Daymond      None      None      None  788.0   
2    None    None       Dan Russell      None      None      None  123.0   
3    None    None   Soneela Nankani      None      None      None  676.0   
4    None    None   Jesse Bernstein      None      None      None  600.0   

  releasedate language  stars  ratings  price  
0  04-08-2008  English    5.0     34.0  468.0  
1  01-05-2018  English    4.5     41.0  820.0  
2  06-11-2020  English    4.5     38.0  410.0  
3  05-10-2021  English    4.5     12.0  615.0  
4  13-01-2010  English    4.5    181.0  820.0

As operações realizadas foram, em resumo:

* **Coluna `name`**: remoção de espaços extras, inclusive antes de '?' e '!'.
* **Coluna `author`**: remoção de espaços extras; remoção da string 'Writtenby:'; adição de espaço entre nome e sobrenome; divisão de autores por coluna.
* **Coluna `narrator`**: remoção de espaços extras; remoção da string 'Narratedby:'; adição de espaço entre nome e sobrenome; divisão de narradores por coluna.
* **Coluna `time`**: remoção de espaços extras; conversão dos valores para minutos; conversão da coluna para `float`.
* **Coluna `releasedate`**: remoção de espaços extras; conversão da coluna para `datetime`; conversão dos valores para o formato dd-mm-aaaa.
* **Coluna `language`**: remoção de espaços extras; capitalização dos valores.
* **Coluna `stars`**: remoção de espaços extras; criação da coluna `ratings` para a qual foi extraído o número de avaliações, convertida para `float`; manutenção apenas da média de avaliações na coluna `stars`, convertida para `float`.
* **Coluna `price`**: remoção de vírgulas; substituição da string 'Free' por 0; conversão da coluna para `float`.

O DataFrame limpo será salvo como arquivo CSV.

In [31]:
df.to_csv('audible-limpo.csv')